File: confd/examples.confd/intro/python/13-stats_push/arpstat.py
"""
*********************************************************************
* Introduction example for pushing data to the CDB oper datastore.  *
* Python version.                                                   *
*                                                                   *
* (C) 2018 Tail-f Systems                                           *
* Permission to use this code as a starting point hereby granted    *
*                                                                   *
* See the README file for more information                          *
*********************************************************************
"""
from __future__ import print_function

import argparse
import logging
import re
import socket
import subprocess
import sys
import textwrap
import time

import _confd
import _confd.dp as dp
import _confd.maapi as maapi

from arpe_ns import ns

CONFD_ADDR = '127.0.0.1'
INTERVAL = 5
MAX_NOBJS = 100

log_level = logging.INFO
logging.basicConfig(
    format="%(asctime)s:%(relativeCreated)s"
           "%(levelname)s:%(filename)s:%(lineno)s:%(funcName)s  %(message)s",
    level=log_level)
log = logging.getLogger("confd_example_logger")
maapisock = None


def make_tag_value(ns_hash, tag, init, vtype):
    """
    Wrapper to create a _confd.TagValue
    """
    return _confd.TagValue(_confd.XmlTag(ns_hash, tag),
                           _confd.Value(init, vtype))


def proccess_arp_line(line):
    log.debug("==>")
    # Now lazy parse lines like
    # ? (192.168.1.1) at 00:0F:B5:EF:11:00 [ether] on eth0
    # slightly different arp output on Linux and BSD

    # Skip space and 'at'
    log.debug("processing line={}".format(line))

    arp_line = [x for x in re.split("[ ,?<>()]", line)
                if x != "" and x != "at"]
    ip = arp_line[0]
    i = 1
    perm = 0
    pub = 0
    hwaddr = ""
    if arp_line[1] == "incomplete":
        for elem in arp_line[1:]:
            i += 1
            if elem == "on":
                break
    elif arp_line[1]:
        hwaddr = arp_line[1]
        for elem in arp_line[2:]:
            i += 1
            if elem == "PERM":
                perm = 1
            elif elem == "PUB":
                pub = 1
            elif elem == "[ether]":
                continue
            elif elem == "on":
                break
    # i should now point on the index just before the interface
    iface = arp_line[i + 1].strip()
    # Some OSes have perm/pub after interface name
    for elem in arp_line[i + 1:]:
        if elem == "permanent":
            perm = 1
        elif elem == "published":
            pub = 1
    log.debug("ip={} hwaddr={} iface={} perm={} pub={}"
              .format(ip, hwaddr, iface, perm, pub))
    v = [
        make_tag_value(ns.hash, ns.arpe_arpe, (ns.arpe_arpe, ns.hash),
                       _confd.C_XMLBEGIN),
        make_tag_value(ns.hash, ns.arpe_ip, ip, _confd.C_IPV4),
        make_tag_value(ns.hash, ns.arpe_ifname, iface, _confd.C_STR),
        make_tag_value(ns.hash, ns.arpe_hwaddr, hwaddr, _confd.C_STR),
        make_tag_value(ns.hash, ns.arpe_permanent, perm, _confd.C_BOOL),
        make_tag_value(ns.hash, ns.arpe_published, pub, _confd.C_BOOL),
        make_tag_value(ns.hash, ns.arpe_arpe, (ns.arpe_arpe, ns.hash),
                       _confd.C_XMLEND),
    ]

    log.debug("<== len(v)={}".format(len(v)))
    return v


def update_arp_data_in_cdb(msock, tid, max_nobjs):
    log.debug("==>")
    v = []

    fp = subprocess.Popen(["arp", "-an"], universal_newlines=True,
                          stdout=subprocess.PIPE)
    maapi.set_namespace(maapisock, tid, ns.hash)
    maapi.delete(maapisock, tid, "/arpentries/arpe")

    for line in iter(fp.stdout.readline, ''):
        v += proccess_arp_line(line)
        if len(v) == max_nobjs + 2:
            log.debug("got {} objects, writing chunk to CDB".format(len(v)))
            maapi.set_values(msock, tid, v, "/arpentries")
            v = []

    if len(v) != 0:
        log.debug("writing objects (len={}) to CDB".format(len(v)))
        maapi.set_values(msock, tid, v, "/arpentries")

    log.debug("<==")


def run_arp(max_nobjs):
    log.debug("==>")

    dbname = _confd.OPERATIONAL
    tid = maapi.start_trans(maapisock, dbname, _confd.READ_WRITE)
    update_arp_data_in_cdb(maapisock, tid, max_nobjs)
    maapi.apply_trans(maapisock, tid, 0)
    maapi.finish_trans(maapisock, tid)

    log.debug("<==")


def init_confd_daemon(confd_debug_level):
    log.info("==>")

    user = "admin"
    groups = [user]
    context = "system"
    global maapisock

    # In C we use confd_init() which sets the debug-level, but for Python the
    # call to confd_init() is done when we do 'import confd'.
    # Therefore we need to set the ConfD debug level here (if we want it to be
    # different from the default debug level - CONFD_SILENT):
    _confd.set_debug(confd_debug_level, sys.stderr)
    dp.init_daemon("arpe_app")
    maapisock = socket.socket()
    maapi.connect(sock=maapisock, ip=CONFD_ADDR, port=_confd.CONFD_PORT)
    maapi.load_schemas(maapisock)  # in Python load schemas through maapi
    maapi.start_user_session(maapisock, user, context, groups, CONFD_ADDR,
                             _confd.PROTO_TCP)

    log.info("<== Initialization complete")


def main():
    log.info("==>")

    debug_levels = {
        's': _confd.SILENT,
        'd': _confd.DEBUG,
        't': _confd.TRACE,
        'p': _confd.PROTO_TRACE,
    }

    rv = _confd.CONFD_OK
    interval = INTERVAL
    max_nobjs = MAX_NOBJS
    try:

        parser = argparse.ArgumentParser(
            description="",
            formatter_class=argparse.RawTextHelpFormatter
        )
        parser.add_argument('-dl', '--debuglevel', choices=debug_levels.keys(),
                            help=textwrap.dedent(
                                '''\
                            set the debug level:
                                s = silent (i.e. no) debug
                                d = debug level debug
                                t = trace level debug
                                p = proto level debug
                            '''))
        parser.add_argument("-i", "--interval", dest="interval",
                            action="store", type=int,
                            help="timeout interval", default=INTERVAL)
        parser.add_argument("-m", "--max_nobj", dest="max_nobjs",
                            action="store", type=int,
                            help="max objs size", default=MAX_NOBJS)
        args = parser.parse_args()
        log.debug('Args = {0}'.format(args))
        if args.interval:
            interval = args.interval
        if args.max_nobjs:
            max_nobjs = args.max_nobjs
        confd_debug_level = debug_levels.get(args.debuglevel, _confd.TRACE)
        log.debug("interval={} max_nobjs={} confd_debug_level={}".
                  format(interval, max_nobjs, confd_debug_level))

        init_confd_daemon(confd_debug_level)
        while True:
            time.sleep(interval)
            run_arp(max_nobjs)
    except KeyboardInterrupt:
        log.info(" **** Ctrl-C pressed ***")
    except Exception:
        log.exception("Error during processing!")
        rv = _confd.CONFD_ERR
    finally:
        pass
        log.debug("Closing sockets")
        if maapisock is not None:
            maapisock.close()
    log.info("<== rv=%d" % rv)
    return rv


if __name__ == "__main__":
    main()

File: confd/examples.confd/intro/13-c_stats_push/arpstat.c
/*********************************************************************
 * Introduction example for pushing data to the CDB oper datastore
 *
 * (C) 2005-2018 Tail-f Systems
 * Permission to use this code as a starting point hereby granted
 * This is ConfD Sample Code.
 *
 * See the README file for more information
 ********************************************************************/

#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <sys/poll.h>
#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <sys/time.h>
#include <time.h>
#include <string.h>
#include <signal.h>

#include <confd_lib.h>
#include <confd_dp.h>
#include <confd_maapi.h>
#include "arpe.h"

/* How often in seconds we read the ARP table into CDB */
#define INTERVAL 5
/* Max chunk of ARP objects / list entries / rows that
   we push into CDB */
#define MAX_NOBJS 100

static int navals;
static int msock, th;

/********************************************************************/
static volatile int do_run_arp = 0;

static void catch_alarm(int sig)
{
    do_run_arp++;
}

/********************************************************************/

/* parse output fom arp -an and push the output to CDB oper */
static void run_arp(int max_nobjs)
{
    char *sep = " ?()<>\n";
    FILE *fp;
    char buf[BUFSIZ];
    confd_tag_value_t *v = (confd_tag_value_t *)
                           malloc(sizeof(confd_tag_value_t) *
                                  (navals+2) *
                                  max_nobjs);
    int i = 0;
    struct in_addr ip4;
    char *hwaddr;
    char *ifname;
    int perm, pub, ret;

    if ((fp = popen("arp -an", "r")) == NULL)
        return;

    if ((th = maapi_start_trans(msock,CONFD_OPERATIONAL,
                                CONFD_READ_WRITE)) < 0) {
        confd_fatal("failed to start trans\n");
    }
    maapi_set_namespace(msock, th, arpe__ns);
    if ((ret = maapi_delete(msock, th, "arpentries/arpe")) < 0) {
        confd_fatal("maapi_delete() failed");
    }

    while (fgets(&buf[0], BUFSIZ, fp) != NULL) {
        char *cp = strtok(&buf[0], sep);
        perm = pub = 0;

        /* Now lazy parse lines like */
        /* ? (192.168.1.1) at 00:0F:B5:EF:11:00 [ether] on eth0 */
        /* slightly different arp output on Linux and BSD */
        ip4.s_addr = inet_addr(cp);
        /* skip "at" */
        assert(strcmp(strtok(NULL, sep), "at") == 0);
        cp = strtok(NULL, sep);

        if ((strcmp(cp, "incomplete") == 0)) {
            assert(strcmp(strtok(NULL, sep), "on") == 0);
            cp = strtok(NULL, sep);
        } else if ((strcmp(cp, "<from_interface>") == 0)) {
            cp = strtok(NULL, sep);
            while (cp) {
                if (strcmp(cp, "on") == 0) {
                    cp = strtok(NULL, sep);
                    break;
                }
                cp = strtok(NULL, sep);
            }
        } else {
            /* some common error cases handled, get real hw addr */
            hwaddr = strdup(cp);

            while (1) {
                cp = strtok(NULL, sep);
                if (cp == NULL)
                    break;
                else if (strcmp(cp, "PERM") == 0)
                    perm = 1;
                else if (strcmp(cp, "PUB") == 0)
                    pub = 1;
                else if (strcmp(cp, "[ether]") == 0)
                    ;
                else if (strcmp(cp, "on") == 0) {
                    cp = strtok(NULL, sep);
                    break;
                }
            }
        }

        /* cp should now point to the interface name
           - this is required since it is a key */
        if (cp) {
            CONFD_SET_TAG_XMLBEGIN(&v[i], arpe_arpe, arpe__ns); i++;
            CONFD_SET_TAG_IPV4(&v[i], arpe_ip, ip4); i++;
            ifname = strdup(cp);
            CONFD_SET_TAG_STR(&v[i], arpe_ifname, ifname); i++;
            CONFD_SET_TAG_STR(&v[i], arpe_hwaddr, hwaddr); i++;

            /* Some OSes have perm/pub after interface name */
            while ((cp = strtok(NULL, sep)) != NULL) {
                if (strcmp(cp, "permanent") == 0)
                    perm = 1;
                else if (strcmp(cp, "published") == 0)
                    pub = 1;
            }

            CONFD_SET_TAG_BOOL(&v[i], arpe_permanent, perm); i++;
            CONFD_SET_TAG_BOOL(&v[i], arpe_published, pub); i++;

            CONFD_SET_TAG_XMLEND(&v[i], arpe_arpe, arpe__ns); i++;
            if(i == max_nobjs*(navals+2)) {
                if ((ret = maapi_set_values(msock,th,v,i,
                                            "/arpentries")) < 0) {
                    confd_fatal("maapi_set_values() failed");
                }
                i = 0;
            }
        } else {
            /* skip this entry */
        }
    }
    if(i > 0) {
        if ((ret = maapi_set_values(msock,th, v, i,
                                    "/arpentries")) < 0) {
            confd_fatal("maapi_set_values() failed");
        }
    }
    if ((ret = maapi_apply_trans(msock, th, 0)) < 0) {
        confd_fatal("maapi_apply_trans() failed");
    }
    maapi_finish_trans(msock, th);
    free(v);
    pclose(fp);
}

/********************************************************************/

int main(int argc, char *argv[])
{
    struct sockaddr_in addr;
    int debuglevel = CONFD_TRACE;
    int interval = INTERVAL;
    int max_nobjs = MAX_NOBJS;
    time_t now;
    struct tm *tm;
    struct itimerval timer;
    struct confd_cs_node *object;
    int c;
    struct confd_ip ip;
    const char *groups[] = { "admin" };
    char *context = "system";

    while ((c = getopt(argc, argv, "i:x:dpts")) != EOF) {
        switch(c) {
        case 'i':
            interval = atoi(optarg);
            break;
        case 'x':
            max_nobjs = atoi(optarg);
            break;
        case 'd':
            debuglevel = CONFD_DEBUG;
            break;
        case 'p':
            debuglevel = CONFD_PROTO_TRACE;
            break;
        case 't':
            debuglevel = CONFD_TRACE;
            break;
        case 's':
            debuglevel = CONFD_SILENT;
            break;
        }
    }

    /* initialize ConfD library */
    confd_init("arpe_app", stderr, debuglevel);

    addr.sin_addr.s_addr = inet_addr("127.0.0.1");
    addr.sin_family = AF_INET;
    addr.sin_port = htons(CONFD_PORT);

    if (confd_load_schemas((struct sockaddr*)&addr,
                           sizeof (struct sockaddr_in)
                          ) != CONFD_OK)
        confd_fatal("failed to load schemas from confd\n");

    object = confd_cs_node_cd(NULL, "/arpe:arpentries/arpe");
    navals = confd_max_object_size(object);

    if ((msock = socket(PF_INET, SOCK_STREAM, 0)) < 0 ) {
        confd_fatal("failed to create socket\n");
    }

    if (maapi_connect(msock, (struct sockaddr*)&addr,
                      sizeof(addr)) < 0) {
        confd_fatal("failed to connect to confd\n");
    }

    ip.af = AF_INET;
    inet_pton(AF_INET, "127.0.0.1", &ip.ip.v4);

    if ((maapi_start_user_session(msock, "admin", context,groups,
                                  sizeof(groups) / sizeof(*groups),
                                  &ip,
                                  CONFD_PROTO_TCP) != CONFD_OK)) {
        confd_fatal("failed to start user session");
    }

    signal(SIGALRM, catch_alarm);

    /* start at next multiple of interval */
    now = time(NULL);
    tm = localtime(&now);
    timer.it_value.tv_sec = interval - tm->tm_sec % interval;
    timer.it_value.tv_usec = 0;
    timer.it_interval.tv_sec = interval;
    timer.it_interval.tv_usec = 0;
    setitimer(ITIMER_REAL, &timer, NULL);

    do_run_arp = 1;

    while(1) {
        pause();
        if (do_run_arp) {
            do_run_arp = 0;
            run_arp(max_nobjs);
        }
    }
}

/********************************************************************/
